### ISSUE TITLE
Suggestion: Improved case .. esac

### ISSUE THREAD

**Issue by dasisdormax** [jfcherng/Sublime-ShellScriptImproved#3]
Currently, inside of a case, everything except strings and keywords is just highlighted as a comment...

What about getting some color in there? :)

Here's some code you can test it on again. I added some more ideas in the comments as well.

``` bash
# More function definitions

longcommand () { echo "Some text"; echo "Some error text" 1>&2; sleep 0.5; return 43; }

longcommand > /dev/null 2>&1

# These should be highlighted equally

# I actually like the redirections having the same color as numbers
# That would make them more visible against the actual commands
( longcommand > /dev/null 2>&1 )
errno=$?

if ( longcommand > /dev/null 2>&1 ); then
    echo "Everything's good!"
fi

# I think highlighting (( .. )) and [[ .. ]] similarly would be consistent
if (( $errno > 42 )); then
    echo "The error is larger than the answer!"
fi

while [[ $1 ]]; do

    case "$1" in

        # Note: the ( is optional
        # The ) and ;; are like then and fi
        # The case names themselves should be highlighted just like strings
        # Inside a case, commands should be commands

        ( info | about | license | "()" | \) )
            info
            ;;

        ( help | --h* | usage )
            usage
            ;;

        ( status )
            if is-running
                then echo "Yes, it is running"
                else echo "No, it is not running"
            fi
            ;;

        ( start )
            if ! is-running; then start; fi
            ;;

        ( stop )
            is-running && stop
            ;;

        ( * )
            echo "ERROR: Unrecognized Option: $1."
            echo "       Try 'some-script usage' for a list of available commands."
            exit 1
            ;;

    esac
    shift

done
```

I hope, my feedback and testing help you to improve ShellScriptImproved even more. Good luck and thanks for your commitment :+1:

**Comment by jfcherng** [comment]
``` bash
 # More function definitions

longcommand () { echo "Some text"; echo "Some error text" 1>&2; sleep 0.5; return 43; }

longcommand > /dev/null 2>&1
```

No sure what you intend to say.

## 

``` bash
# I actually like the redirections having the same color as numbers
# That would make them more visible against the actual commands
```

Yes and I think most people would agree with this. :+1: 
Reference: [I/O Redirection](http://www.tldp.org/LDP/abs/html/io-redirection.html)

## 

``` bash
# I think highlighting (( .. )) and [[ .. ]] similarly would be consistent
if (( $errno > 42 )); then
    echo "The error is larger than the answer!"
fi
```

This repo is not a complete re-write and it starts from the official one.
I am not sure why `((...))` is scoped with `string.other.math.shell` in the official syntax definition hence they are colored like a string.
Reference: [The Double-Parentheses Construct](http://tldp.org/LDP/abs/html/dblparens.html)

## 

``` bash
        # Note: the ( is optional
        # The ) and ;; are like then and fi
        # The case names themselves should be highlighted just like strings
        # Inside a case, commands should be commands
```

No sure what you mean. Is this okay?
![b79cdff4-fe87-11e5-9526-0b249d588288](https://cloud.githubusercontent.com/assets/6594915/14403423/34956792-fe88-11e5-8b17-83c0cbfd41a8.png)
Do you mean the thing inside `(...)` should be scoped or colored as `string`?

**Comment by dasisdormax** [comment]
To answer your questions:

The first two lines are just example code, so `longcommand` is defined. info, usage, and other referenced commands are functions defined in other code.

Regarding `(( .. ))` and `[[ .. ]]`, I was a bit confused that the brackets themselves are colored differently, that's why I included it here. But other than that, I have no problem with `(( .. ))` the way it is right now.

And yes, i think the single cases inside the `(...)` should be colored like strings. I am uncertain right now though, whether things like backticks and variables are processed in there.
In general, a string comparison is happening there, like in `if [[ $1 == info ]]`, and it doesn't matter if you use quotes or not (except if you have spaces inside of the string).

And what i was trying to say with `The ) and ;; are like then and fi`, is that these begin and end the command block inside, so the `;;` should be red as well.

**Comment by jfcherng** [comment]
> The first two lines are just example code, so longcommand is defined. info, usage, and other referenced commands are functions defined in other code.

It's a kind of more powerful static code analyzer which cannot be done with simply `.sublime-syntax` files. Sorry, this won't be fixed.

## 

> Regarding (( .. )) and [[ .. ]], I was a bit confused that the brackets themselves are colored differently, that's why I included it here. But other than that, I have no problem with (( .. )) the way it is right now.

Okay. It makes no sense to me too as it's scoped as a `string`. (fixed in 1.0.14)

## 

> And yes, i think the single cases inside the (...) should be colored like strings. I am uncertain right now though, whether things like backticks and variables are processed in there.
> In general, a string comparison is happening there, like in if [[ $1 == info ]], and it doesn't matter if you use quotes or not (except if you have spaces inside of the string).

Yes, they can be "unquoted" strings but I would prefer not to color/scope them in `case` block. I prefer the screenshot which I provide in my previous comment and that's how VIM deals with them. 

As for in `if` block, at least, it needs a fix for the following code. (fixed in 1.0.14)

``` bash
if [[ $1 == echo ]]; then
    #       ^^^^ this "echo" is a unquoted string
    echo The first argument is "echo"
fi
```

## 

> And what i was trying to say with The ) and ;; are like then and fi, is that these begin and end the command block inside, so the ;; should be red as well.

`;;` is scoped as `punctuation.terminator.case-clause.shell`. So adding something like the following codes to your color scheme would work. I will add this to the readme later.

``` xml
<dict>
    <key>name</key>
    <string>Shell - ;; in "case"</string>
    <key>scope</key>
    <string>punctuation.terminator.case-clause.shell</string>
    <key>settings</key>
    <dict>
        <key>fontStyle</key>
        <string>bold</string>
        <key>foreground</key>
        <string>#F92672</string>
    </dict>
</dict>
```

**Comment by dasisdormax** [comment]
> It's a kind of more powerful static code analyzer which cannot be done with simply .sublime-syntax files. Sorry, this won't be fixed.

I thought so, and didn't request you to. Those two lines really didn't mean anything :)

However, much of the problems I described may be a color scheme issue instead of an issue of your syntax highlighting. With default Solarized Dark, it still looks unchanged to me, but Monokai does it a lot better.

Btw, I really like the way you solved it in your screenshots. Can you tell us the color scheme that you use? :)

**Comment by jfcherng** [comment]
> However, much of the problems I described may be a color scheme issue instead of an issue of your syntax highlighting. With default Solarized Dark, it still looks unchanged to me, but Monokai does it a lot better.

Even using Monokai, there are still some losing colors such as `variables`. But actually `variables` is scoped in the official package. :cry: 

> Btw, I really like the way you solved it in your screenshots. Can you tell us the color scheme that you use? :)

The theme I uses is modified from Monokai Soda.
https://gist.github.com/jfcherng/bf14ed9e1de1a449e62231b1a89a3744

I think adding scopes as mentioned in the readme would work for you. You may found some parts in other languages that looks without highlighting but they actually are scoped. And eventually, you have a customized color scheme...

**Comment by dasisdormax** [comment]
Alright, thanks for the info. With Monokai Soda and the modifications you mention in the README file, everything looks great. (I actually didn't tinker with the theme files before)

Of course, it is not ideal that the different color themes behave that differently sometimes and that some scopes are not implemented at all. But it is probably a really difficult task for color scheme creators as well to consider every scope that will appear in real code and their relations to another.

I don't know if this is currently possible at all, but a "fallback scope" could make sense. Something like: If the color scheme doesn't define scope abc, try rendering as scope xyz.

Other than that, I think that the case .. esac stuff looks good now. I'll tell you if i find something new.

**Comment by jfcherng** [comment]
I just add my dark theme, which is modified from Monokai Soda, to be shipped with 1.1.2.

## 

You can set the syntax specific settings to use the color scheme which is shipped with this package.

To do that,
1. Open a file with `ShellScript Improved` syntax.
2. Go to `Preferences` » `Settings - More` » `Syntax Specific - User`
3. Add the `color_scheme` item into the settings file and then save.

``` javascript
{
    // use a bundled color scheme while writing shell scripts
    "color_scheme": "Packages/ShellScriptImproved/color_schemes/dark.tmTheme"
}
```

> That color scheme is the exact one I use in my Sublime Text. So, there is only a dark one.
> You would get the same highlighting as shown in the screenshot.
